porcupine/plugins/python_venv.py

```diff
@@ -0,0 +1,140 @@
+
+To choose which venv to use, right-click it in directory tree and select
+"Use this Python venv".
+"""
+from __future__ import annotations
+
+import logging
+import shutil
+import sys
+import tkinter
+from functools import partial
+from pathlib import Path
+from typing import Dict
+
+import porcupine.plugins.directory_tree as dirtree
+from porcupine import get_paned_window, settings, utils
+
+log = logging.getLogger(__name__)
+setup_after = ["directory_tree"]
+
+
+def is_venv(path: Path) -> bool:
+    landmarks = [path / "pyvenv.cfg"]
+    if sys.platform == "win32":
+        landmarks.append(path / "Scripts" / "python.exe")
+        landmarks.append(path / "Scripts" / "activate.bat")
+    else:
+        landmarks.append(path / "bin" / "python3")
+        landmarks.append(path / "bin" / "activate")
+    return all(landmark.exists() for landmark in landmarks)
+
+
+def _find_venv(project_root: Path) -> Path | None:
+    # TODO: how well is this all going to work with nested projects, and repos
+    # that contain several subfolders with their own venvs?
+    possible_envs = [path for path in project_root.glob("*env*") if is_venv(path)]
+    if possible_envs:
+        # Pick one consistently. Prefer shorter: env instead of env-old.
+        return min(possible_envs, key=(lambda e: (len(str(e)), e)))
+    log.debug(f"no virtualenvs found in {project_root}")
+    return None
+
+
+def get_venv(project_root: Path) -> Path | None:
+    assert project_root.is_dir()
+    custom_paths: Dict[str, str] = settings.get("python_venvs", Dict[str, str])
+
+    if str(project_root) in custom_paths:
+        from_settings = Path(custom_paths[str(project_root)])
+        if is_venv(from_settings):
+            return from_settings
+        log.warning(f"Python venv is no longer valid: {from_settings}")
+
+    result = _find_venv(project_root)
+    if result is None:
+        log.info(f"No venv found in {project_root}")
+    else:
+        log.info(f"Using {result} as venv of {project_root}")
+        custom_paths[str(project_root)] = str(result)  # Do not automagically switch to new venvs
+        settings.set_("python_venvs", custom_paths)  # custom_paths is copy
+    return result
+
+
+def set_venv(project_root: Path, venv: Path) -> None:
+    assert is_venv(venv), venv
+    custom_paths: dict[str, str] = settings.get("python_venvs", Dict[str, str])
+    custom_paths[str(project_root)] = str(venv)
+    settings.set_("python_venvs", custom_paths)  # custom_paths is copy
+    log.info(f"Venv of {project_root} set to {venv}")
+
+
+# This doesn't use Porcupine's python, unless py or python3 points to it
+def find_python(project_root: Path | None) -> Path | None:
+    if project_root is not None:
+        venv = get_venv(project_root)
+        if venv is not None:
+            log.info(f"Using python from venv: {venv}")
+            if sys.platform == "win32":
+                return venv / "Scripts" / "python.exe"
+            return venv / "bin" / "python3"
```

Comment:
I hate Python 2 but:

```suggestion
            return venv / "bin" / "python"
```